### DISCUSSION TITLE
Would it be feasible to introduce a third AllocationStrategy?

### DISCUSSION THREAD

**Discussion by niekbouman** [capnproto/capnproto#1938] [Ideas]
hi @kentonv

Awesome to see that Cap'n Proto is still alive and kicking after all those (>10) years!

For my use case in a system with a custom allocator that does not like large allocations (where 'large' means 'larger than 128KiB .. 1 MiB or so), I'd like to suggest a new AllocationStrategy for `MallocMessageBuilder` that starts with the heuristic-growth allocation strategy but which is capped by a user-defined limit.
E.g., if the limit is 1 MiB (== 128 * 1024 `capnp::word`s) the first segment would still have the default bytesize (8 KiB), and grow (like the heuristic strategy) until the user defined limit has been reached and from then on behaves like the `FIXED_SIZE strategy`. 
Although I could use the `FIXED_SIZE` AllocationStrategy with 1MiB right from the start, this would of course create overhead in case many `MallocMessageBuilder`s exist simultaneously in memory and most of them are used for small (as in 'few KiB') messages.

Would you welcome a PR to extend the AllocationStrategy enum, and e.g. add another MallocMessageBuilder ctor overload that would have a userMaxAllocSize parameter? (Like FIXED_SIZE, this userMaxAllocSize-limit could still be violated if the user creates a single object whose size exceeds the limit.) 

Or would such a change break existing stuff, or is there some other reason why it would not be a good idea to include it?

I can of course make my own derived class from MallocMessageBuilder with an overridden `allocateSegment` method etc, just like https://github.com/capnproto/capnproto/discussions/1928 , but was wondering if such an extension to the AllocationStrategies could benefit others.

By the way, I found a discussion thread in the old capnproto google-groups:
https://groups.google.com/g/capnproto/c/Tj6Q8aQMuFM/m/gk1Ltk6sLE0J
talking about a max number of segments of 512?  Is there indeed such a hard limit (or can it be altered)?
Does this mean that the max size of a message under the `FIXED_SIZE` allocation strategy is 512 times that fixed size..?

**Comment by niekbouman** [comment]
A simplified version would be to leave the AllocationStrategy enum as is, and pass an additional parameter to the constructor(s), i.e.,
```cpp
  explicit MallocMessageBuilder(uint firstSegmentWords = SUGGESTED_FIRST_SEGMENT_WORDS,
      AllocationStrategy allocationStrategy = SUGGESTED_ALLOCATION_STRATEGY, uint maxAllocSize = unbound(MAX_SEGMENT_WORDS / WORDS)); 
```
introduce an extra class member (initiated by the c'tors):
```cpp
  // ...
  AllocationStrategy allocationStrategy;
  uint maxAllocSize;

  bool ownFirstSegment;
  // ...
```
and slightly alter the implementation of `allocateSegment`:
```cpp
  //...
  } else {
    moreSegments.add(result);
    if (allocationStrategy == capnp::AllocationStrategy::GROW_HEURISTICALLY) {
      // set nextSize = min(nextSize+size, MAX_SEGMENT_WORDS)
      // while protecting against possible overflow of (nextSize+size)
      nextSize = (size <= unbound(maxAllocSize) - nextSize)
        ? nextSize + size : unbound(maxAllocSize);
    }
  }
```
PS is calling `unbound` actually needed in `unbound(maxAllocSize)`? I don't fully grasp what `unbound` does...

**Comment by kentonv** [comment]
Hmm, I think the right thing to do here is for you to create your own subclass of `MessageBuilder`. You can copy the existing code for `MallocMessageBuilder`, and then modify it. This is exactly why `MessageBuilder` is an interface -- so that it is infinitely extensible to any strategy. If lots of people were asking for this, it would maybe make sense to change `MallocMessageBuilder` itself, but you're the first to ask in 10 years, so...

> a max number of segments of 512? Is there indeed such a hard limit (or can it be altered)?

`InputStreamMesasgeReader` enforces such a limit, it order to prevent a malicious client from being able to tell the server to allocate an excessive amount of space for the segment table. But different `MessageReader` implementations could implement different limits, or no limit at all.

I would be OK with extending `ReaderOptions` to make this limit configurable if that's really needed.

**Reply by niekbouman** [reply]
Ok will do that, thanks for your reply! (So far `ReaderOptions` as-is is fine for me)
